src/main/java/com/klodnicki/Bike/v2/service/RentBikeService.java

```diff
@@ -0,0 +1,163 @@
+                    .isRented(bike.isRented())
+                    .bikeType(bike.getBikeType())
+                    .build();
+
+            bikesForNormalUserDTO.add(bikeForNormalDTO);
+
+        }
+        return bikesForNormalUserDTO;
+    }
+
+    @Override
+    public BikeForNormalUserResponseDTO findBikeForNormalUserById(Long id) {
+        Bike bike = bikeService.findBikeById(id);
+        return modelMapper.map(bike, BikeForNormalUserResponseDTO.class);
+    }
+
+    @Override
+    @Transactional
+    public RentResponseDTO rent(RentRequest rentRequest) {
+        Bike bike = bikeService.findBikeById(rentRequest.getBikeId());
+        User user = findUserById(rentRequest.getUserId());
+        ChargingStation chargingStation = findStationById(rentRequest.getChargingStationId());
+
+        bike.setRented(true);
+        bike.setChargingStation(null);
+        bike.setAmountToBePaid(0.0);
+        bike.setUser(user); //it is a relation @OneToOne therefore setting should be on both sides of owning
+        user.setBike(bike);
+        if (chargingStation.getBikeList() != null && !chargingStation.getBikeList().isEmpty()) {
+            chargingStation.getBikeList().remove(bike);
+        }
+        chargingStation.setFreeSlots(chargingStation.getFreeSlots() + 1);
+        int daysOfRent = rentRequest.getDaysOfRent();
+
+        Rent rent = new Rent(LocalDateTime.now(), null, bike, user, null, daysOfRent);
+        rentRepository.save(rent);
+
+        return new RentResponseDTO(
+                rent.getId(),
+                rent.getRentalStartTime(),
+                null,
+                rent.getDaysOfRent(),
+                modelMapper.map(rent, BikeForNormalUserResponseDTO.class),
+                modelMapper.map(rent, UserForNormalUserResponseDTO.class), null);
+    }
+
+    @Override
+    public ChargingStation addBikeToList(Long chargingStationId, Bike bike) {
+        ChargingStation chargingStation = findStationById(chargingStationId);
+        chargingStation.getBikeList().add(bike);
+        //In JPA, only the owning side of the relationship is used when writing to the database.
+        //which means this will not be saved in database. Charging station still will have an empty bike list.
+        //because ChargingStation is NOT the owning-side. Therefore, I must add line below to fix this problem
+        bike.setChargingStation(chargingStation);
+
+        return chargingStationService.save(chargingStation);
+    }
+
+    @Override
+    @Transactional
+    public ResponseEntity<?> returnVehicle(Long rentId, Long returnChargingStationId, Long bikeId) {
+        Bike bike = bikeService.findBikeById(bikeId);
+        ChargingStation returnChargingStation = findStationById(returnChargingStationId);
+        Rent rent = findRentById(rentId);
+        User user = findUserById(bike.getUser().getId());
+
+        bike.setRented(false);
+        bike.setAmountToBePaid(countRentalCost(rentId));
```

Comment:
I believe that the field "amountToBePaid" doesn't fit in Bike class. Fields should describe the object - amount to be paid doesn't do that. The price that must be paid should be placed in the Rent entity. Every rent has a time of rent started and time of rent ended. So it can also hold the calculated value of what is the price on the receipt.
On the vehicle object you can for example hold "picePerMinute" fied if every vehicle in the app will have different values for that. "pricePerDay" could also be a good idea if we want a static values set this way.
But all this depends on the context and business requirements. For sure though this needs further discussion and current implementation has a clear flaw.

Reply:
yes! I agree with that!

Reply:
I have moved field with amount to be paid to Rent entity